janitor/utils.py

```diff
@@ -833,21 +874,54 @@ def _computations_pivot_longer(
         if names_sep is not None:
             between_df = between_df.str.split(names_sep, expand=True)
         else:
-            between_df = between_df.str.extractall(names_pattern).droplevel(-1)
+            # this takes care of scenario 4
+            # and reconfigures it so it takes the scenario 3 path
+            # with `.value`
+            if isinstance(names_pattern, (list, tuple)):
+                condlist = [
+                    between_df.str.contains(regex) for regex in names_pattern
+                ]
+                between_df = np.select(condlist, names_to, None)
+                names_to = [".value"]
+                between_df = pd.DataFrame(between_df, columns=names_to)
+                if between_df.loc[:, ".value"].hasnans:
+                    between_df = between_df.dropna()
+            else:
+                between_df = between_df.str.extractall(
+                    names_pattern
+                ).droplevel(-1)
         # set_axis function labels argument takes only list-like objects
         if isinstance(names_to, str):
             names_to = [names_to]
 
+        # check number of columns
+        # before assigning names_to as `between_df` new columns
         if len(names_to) != between_df.shape[-1]:
             raise ValueError(
                 """
                 Length of ``names_to`` does not match
                 number of columns extracted.
                 """
             )
+
+        # safeguard for a regex that returns nothing
+        if between_df.empty:
+            raise ValueError(
+                """
+                The regular expression in ``names_pattern`` did not
+                return any matches.
+                """
+            )
+
         before_df = _reindex_func(before_df, index)
         between_df = between_df.set_axis(names_to, axis="columns")
 
+        # What should be the appropriate error message
+        # if an inappropriate regular expression is used, that returns
+        # an incorrect number of rows? Where should it be detected?
+        # Personally, I think the onus lies on the user to provide
+        # the right regex.
+
```

Comment:
@samukweku @ericmjl 
Is this still an open topic? If there is a check for the regular expression (i.e. the safeguard check above), I think we are ok here... I agree that the onus should lie on the user to provide the right regex. However, at least some type of check should give the user intuition on why the function could fail. I think this is accomplished with the safeguard check. Just my 2 cents...

Reply:
I'll take this out. Wanted feedback on that (which I've gotten, thanks)